README.md
# SDK PHP PARA API PIX GERENCIANET

Este repositório foi desenvolvido pela Consultoria Técnica Gerencianet e tem o objetivo de facilitar a integração com os serviços oferecidos no contexto do arranjo Pix, como criação de cobrança, verificação de Pix recebidos e devolução, bem como o entendimento do qrcode, que deve ser construído seguindo a especificação do [BrCode](https://www.bcb.gov.br/content/estabilidadefinanceira/spb_docs/ManualBRCode.pdf).

## Requisitos
* PHP >= 5.4


## Testado com as versões
```
php 5.4, php 7.0 e php 7.3
```

## Como usar?
Para começar, você deve configurar os parâmetros no arquivo config.json.
Instancie as informações usando sua `chave PIX` vinculada à sua conta Gerencianet, `client_id`, `client_secret` da sua aplicação e `sandbox` igual a *true*, caso seu ambiente seja de Homologação, ou *false*, caso seja Produção.

```json
{
    "sandbox": true,
    "chave": "chavepix@email.com",
    "homologacao": {
        "nome_certificado": "developmentCertificate.pem",
        "client_id": "Client_Id_...",
        "client_secret": "Client_Secret_...",
        "pix_url_auth": "https://api-pix-h.gerencianet.com.br/oauth/token",
        "pix_url_cob": "https://api-pix-h.gerencianet.com.br/v2/cob",
        "pix_url": "https://api-pix-h.gerencianet.com.br/v2/pix"
    },
    "producao": {
        "nome_certificado": "productionCertificate.pem",
        "client_id": "Client_Id_...",
        "client_secret": "Client_Secret_...",
        "pix_url_auth": "https://api-pix.gerencianet.com.br/oauth/token",
        "pix_url_cob": "https://api-pix.gerencianet.com.br/v2/cob",
        "pix_url": "https://api-pix.gerencianet.com.br/v2/pix"
    }
}
```

Todas as requisições devem conter um certificado de segurança que será fornecido pela Gerencianet dentro da sua conta. 

Para gerar o seu certificado abra um ticket em https://gerencianet.com.br/fale-conosco informando o numero de sua conta e as credenciais client_id e client_secret de desenvolvimento. Nossa equipe irá retornar com o certificado `.p12` para você realizar o consumo dos endpoints.

Para a utilização no PHP o certificado deverá ser convertido em `.pem`. Segue exemplos utilizando o OpenSSL para a conversão.

**Comando 1**
```
// Gerar certificado e chave em único arquivo
openssl pkcs12 -in certificado.p12 -out certificado.pem -nodes
```

**Comando 2**
```
// Gerar certificado e chave separadas
openssl pkcs12 -in path.p12 -out newfile.crt.pem -clcerts -nokeys #certificado
openssl pkcs12 -in path.p12 -out newfile.key.pem -nocerts -nodes #chave privada
```

Após a conversão dos certificados, salvá-los no diretório `./certificado` com os mesmos nomes contidos no `config.json`.


## Exemplos de execução
Você pode executar usando qualquer servidor web, como Apache ou nginx, ou simplesmente iniciar um servidor php da seguinte forma:

```php
php -S localhost:9000
```

Em seguida, abra qualquer exemplo em seu navegador.

:warning: Alguns exemplos requerem que você altere alguns parâmetros para funcionar, como por exemplo, o arquivo `emitirPix.php` onde você deve inserir os parâmetros da cobrança.


## Documentação adicional

A documentação completa com todos os endpoints disponíveis está em https://dev.gerencianet.com.br/docs.


## Licença ##
Este projeto está licenciado sob a Licença MIT - consulte o arquivo [LICENSE.md](LICENSE) para obter detalhes


CHANGELOG.md
## [1.0.0] - 2020-12-09

### Added

- Field "recebedor" in the "config.json" and in the "emitirPix.php" for mount the Qr Code.
- Generate txID for "solicitaDevolucaoPix.php".
 
### Changed
- Fix function `getTxID()`
- Added field "recebedor"  in the config README example.
- Case QR Code "dinamico" the txID receive `***` in the  "montaBrCode.php".
 
### Removed

- Verification of the "payload" when assembling the Qr Code.

# [0.1.0] - 2020-11-26

- Initial release.


funcoes.php
<?php

/*
 * Este arquivo foi implementado pela Consultoria Técnica Gerencianet.
 * Data da versão do arquivo: 26/11/2020.
 *
 * Link para documentação técnica da API-Pix da Gerencianet: https://gerencianet.com.br/api-pix
 * Você pode entrar em contato conosco sempre que precisar.
 * Visite o site da Gerencianet: https://www.gerencianet.com.br
*/

function getTxID($tipo)
{
  /*
   # Esta rotina gera um txID de forma aleatória, atendendo à regex `[a-zA-Z0-9]{26,35}$` caso o QR Code seja dinâmico, e à regex [a-zA-Z0-9]{1,25} caso seja estático, conforme padrão bacen
   */

  $id = ($tipo === "devolucao") ? "D" : "C"; // Se o tipo for devolução, o ID iniciará com "D", se não iniciará com "C" de cobrança
  $quantidade = ($tipo === "dinamico") ? 34 : 24;

  for (
    $id .= '', $i = 0, $z = strlen($a = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789') - 1;
    $i != $quantidade;
    $x = rand(0, $z), $id .= $a{
      $x}, $i++
  );
  return $id;
}


function emitePix($pix_url_cob, $arq_certificado, $body, $tokenType, $accessToken)
{
  /*
* Esta rotina consome um endpoid PUT da Gerencianet para emissão da cobrança
*/
  $curl = curl_init();

  curl_setopt_array($curl, array(
    CURLOPT_URL => $pix_url_cob,
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_ENCODING => "",
    CURLOPT_MAXREDIRS => 10,
    CURLOPT_TIMEOUT => 0,
    CURLOPT_FOLLOWLOCATION => true,
    CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
    CURLOPT_CUSTOMREQUEST => "PUT",
    CURLOPT_SSLCERT => $arq_certificado,
    CURLOPT_SSLCERTPASSWD => "",
    CURLOPT_POSTFIELDS => $body,
    CURLOPT_HTTPHEADER => array(
      "authorization: $tokenType $accessToken",
      "Content-Type: application/json"
    ),
  ));

  $dadosPix = json_decode(curl_exec($curl), true);
  curl_close($curl);

  return $dadosPix;
}


function getAccessToken($pix_url_auth, $arq_certificado, $client_id, $client_secret)
{
  /*
   # Esta rotina consome um endpoid POST da Gerencianet para realizar a geração do AccessToken
   */

  $curl = curl_init();

  $authorization =  base64_encode("$client_id:$client_secret");

  curl_setopt_array($curl, array(
    CURLOPT_URL => $pix_url_auth, // Rota base, desenvolvimento ou produção
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_ENCODING => "",
    CURLOPT_MAXREDIRS => 10,
    CURLOPT_TIMEOUT => 0,
    CURLOPT_FOLLOWLOCATION => true,
    CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
    CURLOPT_CUSTOMREQUEST => "POST",
    CURLOPT_POSTFIELDS => '{"grant_type": "client_credentials"}',
    CURLOPT_SSLCERT => $arq_certificado, // Caminho do certificado
    CURLOPT_SSLCERTPASSWD => "",
    CURLOPT_HTTPHEADER => array(
      "Authorization: Basic $authorization",
      "Content-Type: application/json"
    ),
  ));

  $response = curl_exec($curl);

  curl_close($curl);

  return json_decode($response, true);
}


function getPayload($location)
{
  /*
   # Esta rotina consome a URL location da cobrança buscando as informações do JWS
   */
  $curl = curl_init();

  curl_setopt_array($curl, array(
    CURLOPT_URL => $location,
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_ENCODING => "",
    CURLOPT_MAXREDIRS => 10,
    CURLOPT_TIMEOUT => 0,
    CURLOPT_FOLLOWLOCATION => true,
    CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
    CURLOPT_CUSTOMREQUEST => "GET",
    CURLOPT_USERAGENT => "Mozilla/5.0 (Windows NT 10.0;Win64;x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36"
  ));

  $objectJWS = curl_exec($curl);
  curl_close($curl);

  if (is_array(json_decode($objectJWS, true))) { // Se no retorno apresentar erro, será um array, então retornará a mensagem de erro ocorrida
    $message = json_decode($objectJWS, true);
    return $message;
  } else {
    $objectJWS = explode(".", $objectJWS);
    return json_decode(base64_decode($objectJWS[1]), true); // Obtendo suesso, retornará as informações da cobrança
  }
}

function preencheCampo($valor)
{
  /*
   * Esta função retorna a string preenchendo com 0 na esquerda, com tamanho o especificado, concatenando com o valor do campo
   */
  return str_pad(strlen($valor), 2, '0', STR_PAD_LEFT) . $valor;
}


function calculaChecksum($str)
{
  /*
   * Esta função auxiliar calcula o CRC-16/CCITT-FALSE
   */

  function charCodeAt($str, $i)
  {
    return ord(substr($str, $i, 1));
  }

  $crc = 0xFFFF;
  $strlen = strlen($str);
  for ($c = 0; $c < $strlen; $c++) {
    $crc ^= charCodeAt($str, $c) << 8;
    for ($i = 0; $i < 8; $i++) {
      if ($crc & 0x8000) {
        $crc = ($crc << 1) ^ 0x1021;
      } else {
        $crc = $crc << 1;
      }
    }
  }
  $hex = $crc & 0xFFFF;
  $hex = dechex($hex);
  $hex = strtoupper($hex);

  return $hex;
}

function gerarQrCode($payloadBrCode, $tamanhoQrCode)
{
  /*
   * Esta rotina consome uma api da Gerencianet que gera o QR Code
   * 
   * Link para Repositório da API: https://github.com/ceciliadeveza/gerarqrcodepix
   */
  $curl = curl_init();

  curl_setopt_array($curl, array(
    CURLOPT_URL => "https://gerarqrcodepix.com.br/api/v1?brcode=$payloadBrCode&tamanho=$tamanhoQrCode",
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_ENCODING => '',
    CURLOPT_MAXREDIRS => 10,
    CURLOPT_TIMEOUT => 0,
    CURLOPT_FOLLOWLOCATION => true,
    CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
    CURLOPT_CUSTOMREQUEST => 'GET',
  ));

  $imagemQrCode = curl_exec($curl);

  curl_close($curl);

  return base64_encode($imagemQrCode);
}


function verficaFalhas($dadosPix)
{
  if (isset($dadosPix["mensagem"]) || !$dadosPix) { // Se apresentar falha, irá mostrar a mensagem
    echo (!$dadosPix) ? "Falha ao gerar a cobrança, tente novamente" : "Falha PIX<br>Mensagem: " . $dadosPix["mensagem"];

    if (isset($dadosPix["erros"])) { // Se apresentar erro, irá mostrar o erro e o campo
      foreach ($dadosPix["erros"] as $key => $erro) {
        echo "<br><br>Erro " . ($key + 1) . "<br>Caminho: " . $erro["caminho"] . "<br>Mensagem de erro: " . $erro["mensagem"] . "<br><br>";
      }
    }
    exit;
  }
}


function mapped_implode($glue, $array, $symbol = '=')
{
  return implode($glue, array_map(
    function ($k, $v) use ($symbol) {
      return $k . $symbol . $v;
    },
    array_keys($array),
    array_values($array)
  ));
}


montaBrCode.php
<?php

/*
 * Este arquivo foi implementado pela Consultoria Técnica Gerencianet.
 * Data da versão do arquivo: 26/11/2020.
 *
 * Link para documentação técnica da API-Pix da Gerencianet: https://gerencianet.com.br/api-pix
 * Você pode entrar em contato conosco sempre que precisar.
 * Visite o site da Gerencianet: https://www.gerencianet.com.br
*/

function montaBrCode($dadosPix, $tipo, $pagoUmaVez, $cidade, $cep, $valorLivre, $tamanhoQrCode)
{

  $payload = getPayload($dadosPix["loc"]["location"]);

  if (!isset($payload["mensagem"])) { // Se o retorno da consulta do payload apresentar erro, não irá montar o QR Code e mostrará a mensagem

    // Rotina montará a variável que correspondente ao payload no padrão EMV-QRCPS-MPM

    $payload_format_indicator = '01';
    $point_of_initiation_method = '12';
    $merchant_account_information = '00' . preencheCampo('BR.GOV.BCB.PIX');
    $merchant_category_code = '0000';
    $transaction_currency = '986';
    $country_code = 'BR';


    $payloadBrCode = "00" . preencheCampo($payload_format_indicator); // [obrigatório] Payload Format Indicator, valor fixo: 01

    if ($pagoUmaVez) { // Se o QR Code for para pagamento único (só puder ser utilizado uma vez), a variável $pagoUmaVez deverá ser true
      $payloadBrCode .= "01" . preencheCampo($point_of_initiation_method); // [opcional] Point of Initiation Method Se o valor 12 estiver presente, significa que o BR Code só pode ser utilizado uma vez.
    }

    if ($tipo === "dinamico") {
      $location = str_replace("https://", "", $dadosPix["loc"]["location"]); // URL payload do PSP do recebedor que contém as informações da cobrança
      $merchant_account_information .= '25' . preencheCampo($location);
    } else { // Caso seja estático
      $merchant_account_information .= "01" . preencheCampo($dadosPix["chave"]); //Chave do destinatário do pix, pode ser EVP, e-mail, CPF ou CNPJ.
    }
    $payloadBrCode .= "26" .  preencheCampo($merchant_account_information); // [obrigatório] Indica arranjo específico; “00” (GUI) e valor fixo: br.gov.bcb.pix

    $payloadBrCode .= '52' . preencheCampo($merchant_category_code); // [obrigatório] Merchant Category Code “0000” ou MCC ISO18245

    $payloadBrCode .= '53' . preencheCampo($transaction_currency); // [obrigatório] Moeda, “986” = BRL: real brasileiro - ISO4217

    $payloadBrCode .= '54';  // [opcional] Valor da transação. Utilizar o . como separador decimal.
    $payloadBrCode .= ($valorLivre === true) ? preencheCampo('0') : ($tipo === "dinamico") ? preencheCampo($payload["valor"]["original"]) : preencheCampo($dadosPix["valor"]["original"]);

    $payloadBrCode .= '58' . preencheCampo($country_code); // [obrigatório] “BR” – Código de país ISO3166-1 alpha 2

    $payloadBrCode .= '59'; // [opcional] Nome do beneficiário/recebedor. Máximo: 25 caracteres.
    $payloadBrCode .= ($tipo === "dinamico") ? preencheCampo($payload["devedor"]["nome"]) : preencheCampo($dadosPix["devedor"]["nome"]); // [opcional] Nome do beneficiário/recebedor. Máximo: 25 caracteres.

    $payloadBrCode .= '60' . preencheCampo($cidade); // [obrigatório] Nome cidade onde é efetuada a transação. Máximo 15 caracteres.

    $payloadBrCode .= '61' . preencheCampo($cep); // [opcional] CEP da cidade onde é efetuada a transação.

    $txID = ($tipo === "dinamico") ? $payload["txid"] : $dadosPix["txid"];
    $aditional_data_field_template = '05' . preencheCampo($txID); // [opcional] Identificador da transação.
    $payloadBrCode .= '62' . preencheCampo($aditional_data_field_template);


    $payloadBrCode .= "6304"; // Adiciona o campo do CRC no fim da linha do pix.

    $payloadBrCode .= calculaChecksum($payloadBrCode); // Calcula o checksum CRC16 e acrescenta ao final.


    echo "Pix Copia e Cola: <input readonly='readonly' id='inputTest' value='$payloadBrCode' /> <button onclick='copiarTexto()''>Copiar</button> <br>";
    echo "<script>
              let copiarTexto = () =>{
                  //captura o elemento input
                  const inputTest = document.querySelector('#inputTest');
                  
                  //seleciona todo o texto do elemento
                  inputTest.select();
                  //executa o comando copy
                  //aqui é feito o ato de copiar para a area de trabalho com base na seleção
                  document.execCommand('copy');
              };
          </script>";

    $imageString = gerarQrCode($payloadBrCode, $tamanhoQrCode);

    // Exibe a imagem diretamente no navegador codificada em base64.
    return '<img src="data:image/png;base64,' . $imageString . '"></center>';
  } else {
    return '<p>' . $payload["mensagem"] . '</p>';
  }
}


solicitaDevolucaoPix.php
<?php

/*
 * Este arquivo foi implementado pela Consultoria Técnica Gerencianet.
 * Data da versão do arquivo: 26/11/2020.
 *
 * Link para documentação técnica da API-Pix da Gerencianet: https://gerencianet.com.br/api-pix
 * Você pode entrar em contato conosco sempre que precisar.
 * Visite o site da Gerencianet: https://www.gerencianet.com.br
*/

require "./funcoes.php";

/**
 * Representa uma solicitação de devolução de um Pix realizado, cujos fundos já se encontrem disponíveis na conta transacional do usuário recebedor
 * Estados da devolução:
 * a) EM_PROCESSAMENTO: indica que a devolução foi solicitada, mas ainda está em processamento no SPI;
 * b) DEVOLVIDO: indica que a devolução foi liquidada pelo SPI; e
 * c) NAO_REALIZADO: indica que a devolução não pode ser realizada em função de algum erro durante a liquidação (exemplo: saldo insuficiente).
 */

// Insira o identificador da transação (txID) que deseja solicitar a devolução
$txID = "775ASPFF2GW0L63KJ6LHS5RZF0MC0S733WD";

// EndToEndId da transação (e2eid) que deseja consultar. Obs: O e2eid é gerado em cobranças que foram quitadas
$e2eid = "E9040088820201121002700006627428";

// Insira no arquivo config.json suas credenciais e o nome do seu arquivo de certificado .pem
$arquivo_config = file_get_contents("./config.json");
$configuracoes = json_decode($arquivo_config, true);
$ambiente = ($configuracoes["sandbox"] === true) ? "homologacao" : "producao"; // Define o ambiente de acordo com o sadbox em configurações

$arq_certificado = "./certificado/" . $configuracoes[$ambiente]["nome_certificado"]; // Local onde está salvo o certificado .pem

// Busca informações do token de autenticação de acordo com suas credencias e certificado
$dadosToken = getAccessToken($configuracoes[$ambiente]["pix_url_auth"], $arq_certificado, $configuracoes[$ambiente]["client_id"], $configuracoes[$ambiente]["client_secret"]);
$tokenType = $dadosToken['token_type'];
$accessToken = $dadosToken['access_token'];

$pix_url = $configuracoes[$ambiente]["pix_url"] . "/$e2eid/devolucao/$txID"; // Monta a url para a requisição de solicitação de devolução

// Insira as informações que deseja atualizar. Comentar a linha caso não deseja atualizar algum campo
$body = json_encode([
    "valor" =>  "0.01"
]);


$curl = curl_init();

curl_setopt_array($curl, array(
    CURLOPT_URL => $pix_url,
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_ENCODING => "",
    CURLOPT_MAXREDIRS => 10,
    CURLOPT_TIMEOUT => 0,
    CURLOPT_FOLLOWLOCATION => true,
    CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
    CURLOPT_CUSTOMREQUEST => "PUT",
    CURLOPT_SSLCERT => $arq_certificado,
    CURLOPT_SSLCERTPASSWD => "",
    CURLOPT_POSTFIELDS => $body,
    CURLOPT_HTTPHEADER => array(
        "authorization: $tokenType $accessToken",
        "Content-Type: application/json"
    ),
));

$dadosPix = json_decode(curl_exec($curl), true);
curl_close($curl);

verficaFalhas($dadosPix); // Se encontrar falhas, apresentará a mensagem de erro e encerrará a execução

echo "<pre>";
var_dump($dadosPix);
echo "</pre>";


emitirPix.php
<?php

/*
 * Este arquivo foi implementado pela Consultoria Técnica Gerencianet.
 * Data da versão do arquivo: 26/11/2020.
 *
 * Link para documentação técnica da API-Pix da Gerencianet: https://gerencianet.com.br/api-pix
 * Você pode entrar em contato conosco sempre que precisar.
 * Visite o site da Gerencianet: https://www.gerencianet.com.br
*/

require "./funcoes.php";
require "./montaBrCode.php";

$tipo = "dinamico"; // Insira o tipo de QR Code que deseja gerar. Opções: dinamico ou estatico

$tamanhoQrCode = "250"; // Insira o tamanho em pixels do QR Code que será gerado. Obs: Tamanho deve ser m sou igual a 256px

$pagoUmaVez = false; // Insira se deseja que o QR Code pode ser pago uma única vez. Opções: true ou false

$valorLivre = false; // Insira se deseja que o valor da cobrança seja livre. Opções: true ou false

$cidade = "Ouro Preto"; // [obrigatório] cidade onde é efetuada a transação

$cep = "35400000"; // [opcional] CEP da localidade onde é efetuada a transação


// Insira no arquivo config.json suas credenciais e o nome do seu arquivo de certificado .pem
$arquivo_config = file_get_contents("./config.json");
$configuracoes = json_decode($arquivo_config, true);
$ambiente = ($configuracoes["sandbox"] === true) ? "homologacao" : "producao"; // Define o ambiente de acordo com o sadbox em configurações

$arq_certificado = "./certificado/" . $configuracoes[$ambiente]["nome_certificado"]; // Local onde está salvo o certificado .pem

// Busca informações do token de autenticação de acordo com suas credencias e certificado
$dadosToken = getAccessToken($configuracoes[$ambiente]["pix_url_auth"], $arq_certificado, $configuracoes[$ambiente]["client_id"], $configuracoes[$ambiente]["client_secret"]);
$tokenType = $dadosToken['token_type'];
$accessToken = $dadosToken['access_token'];

$txID = getTxID($tipo); // Função gera um txID aleatório obedecendo a regex de cada tipo de QR Code
$pix_url_cob = $configuracoes[$ambiente]["pix_url_cob"] . "/" . $txID; // Monta a url para a requisição que gera a cobrança

// Insira as informações da cobrança
$body = json_encode([
  "calendario" => [
    "expiracao" => 3600 // [opcional] Tempo de vida da cobrança, especificado em segundos a partir da data de criação. Caso não definido o padrão será de 86400 segundos ( 24 horas)
  ],
  "devedor" => [
    "cpf" => "12345678900", // [opcional] CPF do usuário pagador.string /^\d{11}$/
    //"cnpj" => "12345678000123",
    "nome" => "Nome Aqui" // [opcional] Nome do usuário pagador. string (Nome) ≤ 200 characters
  ],
  "valor" => [
    "original" => "1.49" // [obrigatório/opcional] Valor original da cobrança.string \d{1,10}.\d{2} Obs: Na cobrança para QR Code dinâmico é obrigatório um valor, para estático não
  ],
  "chave" => $configuracoes["chave"], // [obrigatório] Determina a chave Pix registrada no DICT que será utilizada para a cobrança.
  "solicitacaoPagador" => "Serviço realizado.", // [opcional] determina um texto a ser apresentado ao pagador para que ele possa digitar uma informação correlata, em formato livre, a ser enviada ao recebedor.
  "infoAdicionais" => [ // [opcional] Cada respectiva informação adicional contida na lista (nome e valor) deve ser apresentada ao pagador.
    [
      "nome" => "Campo 1", // Nome do campo string (Nome) ≤ 50 characters
      "valor" => "Informação Adicional1 do PSP-Recebedor" // Dados do campo string (Valor) ≤ 200 characters
    ],
    [
      "nome" => "Campo 2",
      "valor" => "Informação Adicional2 do PSP-Recebedor"
    ]
  ]
]);


if ($tipo === "dinamico") { // Somente será gerado o payload para o QR Code dinâmico
  $dadosPix = emitePix($pix_url_cob, $arq_certificado, $body, $tokenType, $accessToken);

  verficaFalhas($dadosPix); // Se encontrar falhas, apresentará a mensagem de erro e encerrará a execução
} else { // Caso seja estático enviará as informações contidas no $body
  $dadosPix = json_decode($body, true);
  $dadosPix["txid"] = $txID;
}

$imagemQrCode = montaBrCode($dadosPix, $tipo, $pagoUmaVez, $cidade, $cep, $valorLivre, $tamanhoQrCode); // Monta o pix Copia e Cola e a imagem do QR Code

echo $imagemQrCode;
